File: Retos/Reto #21 - NÚMEROS PRIMOS GEMELOS [Media]/ruby/test0n3.rb
# frozen_string_literal: true

# class for Prime Numbers
class PrimeNumber
  def initialize(limit)
    @limit = limit
  end

  def twin_primes
    temp = []
    twin_primes = []
    return "#{@limit} is not valid input" if !(@limit.is_a? Integer) || (@limit < 2)

    (2..@limit).to_a.each do |number|
      next unless prime?(number)

      temp.push(number)
      next unless temp.size == 2

      a, b = temp
      twin_primes.push([a, b]) if b - a == 2
      temp.shift
    end
    twin_primes
  end

  private

  def prime?(number)
    limit = Integer.sqrt(number)
    limit = number - 1 if limit <= 2
    (2..limit).none? { |divisor| (number % divisor).zero? }
  end
end

# puts "twin_primes: #{PrimeNumber.new(30).twin_primes}"

File: Retos/Reto #21 - NÚMEROS PRIMOS GEMELOS [Media]/javascript/test0n3.js
function PrimeNumber(input) {
  let limit = input;
  let twinPrimes = setTwinPrimes();

  function validLimit(input) {
    if (input < 2) {
      return false;
    }
    return true;
  }

  function isPrime(num) {
    let limit = Math.floor(Math.sqrt(num));
    if (limit <= 2) {
      limit = num - 1;
    }
    for (let divisor = 2; divisor <= limit; divisor++) {
      if (num % divisor == 0) {
        return false;
      }
    }
    return true;
  }

  function setTwinPrimes() {
    let temp = [];
    let resp = [];

    if (!validLimit(limit)) {
      return `${input} is not a valid input`;
    }

    for (let number = 2; number <= limit; number++) {
      if (!isPrime(number)) {
        continue;
      }

      temp.push(number);

      if (temp.length < 2) {
        continue;
      }

      let a = temp[0];
      let b = temp[1];
      if (b - a == 2) {
        resp.push([a, b]);
      }
      temp.shift();
    }
    return resp;
  }

  Object.defineProperty(this, "twinPrimes", {
    get: function () {
      return twinPrimes;
    },
  });
}

// console.log(new PrimeNumber(1).twinPrimes);
// console.log(new PrimeNumber(3).twinPrimes);
// console.log(new PrimeNumber(14).twinPrimes);
